### DISCUSSION TITLE
Issues with the "Fields" Handler and Get-PnPSiteTemplate

### DISCUSSION THREAD

**Discussion by renewtx** [pnp/powershell#4628] [Q&A] [answered]
Wondering if anyone else has encountered a similar issue or might know what the issue is ... I am using **Get-PnPSIteTemplate** to extract templates files for various site components from an existing M365 SharePoint site.  The site was created using a standard M365 Team Site Template (this is NOT a TEAMS-connected site).

I am using PowerShell v7.4.5 with the PnP.PowerShell module v2.99.35-nightly (so relatively recent)

Using Get-PnPSiteTemplate I can extract XML template files for the majority of the handlers we need.  For example, all of the following handlers (and several others) work fine:
- WebSettings
- PropertyBagEntries
- Features
- SearchSettings
- Publishing
- SiteSecurity
- ContentTypes
- Lists

The one handler that does not seem to work is the "Fields" handler (which we have used successfully in the past with earlier versions of PowerShell and PnP.PowersShell, both on-prem and with SPO).  

Currently when I run the extract for the Fields handler as shown below

        **$handlers = "Fields"
        Get-PnPSiteTemplate -Out  "C:\Temp\Template\Template-2-SPFIELD.xml" -Handlers $handlers -ErrorAction Continue -Verbose**

the cmdlet begins to run and then seems to begin 'extracting', but then stops almost immediately after.  I get the following output on the command line:

PS C:\Temp> Get-PnPSiteTemplate -Out  "C:\Temp\Template\Template-2-SPFIELD.xml" -Handlers $handlers -ErrorAction Continue -Verbose
VERBOSE: Making call to SharePoint Online using the Client Side Object Model (CSOM)
VERBOSE: Evaluating application permissions in access token for audience SharePoint Online
VERBOSE: Access token contains the following 2 application permission scopes for resource SharePoint Online: User.ReadWrite.All, Sites.FullControl.All
VERBOSE: No required permissions have been defined on this cmdlet
**Get-PnPSiteTemplate: Object reference not set to an instance of an object.**

In the Debug log file that is generated there is nothing that offers any clues as to what the issue might be ...

pwsh Information: 0 : 2024-12-09 18:14:45.9192	[GetContextAsync]	[0]	[Debug]	Authentication type: AzureADCertificate for scopes https://tenant.sharepoint.com/.default	0ms	
pwsh Information: 0 : 2024-12-09 18:14:46.3545	[PnP.Framework]	[0]	[Debug]	Cloning context for https://tenant.sharepoint.com/sites/KEEPitTemplateSiteRel2.0	0ms	
pwsh Information: 0 : 2024-12-09 18:14:46.3550	[PnP.Framework]	[0]	[Debug]	Checking for different audience https://tenant.sharepoint.com/sites/KEEPitTemplateSiteRel2.0	0ms	
pwsh Information: 0 : 2024-12-09 18:14:46.3709	[GetAccessTokenAsync]	[0]	[Debug]	Authentication type: AzureADCertificate	0ms	
pwsh Information: 0 : 2024-12-09 18:14:47.4038	[Extraction]	[18]	[Debug]	Code execution scope started	0ms	9be3dbfb-e7fa-4dd0-b026-b501b6b860e1
pwsh Information: 0 : 2024-12-09 18:14:47.4040	[Extraction]	[18]	[Debug]	Base template available: GROUP0template	0ms	9be3dbfb-e7fa-4dd0-b026-b501b6b860e1
pwsh Information: 0 : 2024-12-09 18:14:47.4041	[Extraction]	[18]	[Debug]	ProgressDelegate registered	0ms	9be3dbfb-e7fa-4dd0-b026-b501b6b860e1
pwsh Information: 0 : 2024-12-09 18:14:47.4041	[Extraction]	[18]	[Debug]	MessagesDelegate registered	0ms	9be3dbfb-e7fa-4dd0-b026-b501b6b860e1
pwsh Information: 0 : 2024-12-09 18:14:47.4087	[PnP.Framework]	[0]	[Debug]	Cloning context for https://tenant.sharepoint.com/sites/KEEPitTemplateSiteRel2.0	0ms	
pwsh Information: 0 : 2024-12-09 18:14:47.4088	[PnP.Framework]	[0]	[Debug]	Checking for different audience https://tenant.sharepoint.com/sites/KEEPitTemplateSiteRel2.0	0ms	
pwsh Information: 0 : 2024-12-09 18:14:47.4117	[Fields]	[18]	[Debug]	Code execution scope started	1ms	9be3dbfb-e7fa-4dd0-b026-b501b6b860e1
pwsh Information: 0 : 2024-12-09 18:14:47.9430	[Fields]	[18]	[Debug]	Code execution scope ended	532ms	9be3dbfb-e7fa-4dd0-b026-b501b6b860e1
pwsh Information: 0 : 2024-12-09 18:14:47.9436	[Extraction]	[18]	[Debug]	Code execution scope ended	539ms	9be3dbfb-e7fa-4dd0-b026-b501b6b860e1

Anyone else used the Fields handler successfully?  Ideas as to why this is not working?

Thanks

**Comment by waaromikniet** [comment, marked as answer]
I tried your command and get no errors.
Do you have all required application permissions granted for your Azure AD app?
Do you have taxonomy fields in your site? Maybe need termstore permissions?

**Comment by renewtx** [comment]
Hi,

Thanks for responding ... yes, I believe all the relevant permissions have been granted.  There are some taxonomy fields but the app has those necessary permissions as well 

![image](https://github.com/user-attachments/assets/e75fb61c-ff1b-431d-99c1-1a2dbddccc6c)

When I run the cmdlet (after connecting) I just get this output:

Get-PnPSiteTemplate: **Object reference not set to an instance of an object.**

Any ideas how to get more details?  The only thing that shows up in the log file is:

powershell_ise.exe Information: 0 : 2024-12-11 10:58:59.5264	[Site Security]	[19]	[Debug]	Processing Role Assignment Microsoft.SharePoint.Client.RoleAssignment	772ms	d0bff343-f822-4ca0-aa1f-88864a38a900
powershell_ise.exe Information: 0 : 2024-12-11 10:58:59.5264	[Site Security]	[19]	[Debug]	Processing Role Assignment Microsoft.SharePoint.Client.RoleAssignment	772ms	d0bff343-f822-4ca0-aa1f-88864a38a900
powershell_ise.exe Information: 0 : 2024-12-11 10:58:59.5317	[Site Security]	[19]	[Debug]	Code execution scope ended	776ms	d0bff343-f822-4ca0-aa1f-88864a38a900
powershell_ise.exe Information: 0 : 2024-12-11 10:58:59.5406	[Fields]	[19]	[Debug]	Code execution scope started	0ms	d0bff343-f822-4ca0-aa1f-88864a38a900
powershell_ise.exe Information: 0 : 2024-12-11 10:59:00.2391	[Fields]	[19]	[Debug]	Code execution scope ended	698ms	d0bff343-f822-4ca0-aa1f-88864a38a900
powershell_ise.exe Information: 0 : 2024-12-11 10:59:00.2391	[Extraction]	[19]	[Debug]	Code execution scope ended	3293ms	d0bff343-f822-4ca0-aa1f-88864a38a900

**Reply by waaromikniet** [reply]
Do you see more when you run 

`Set-PnPTraceLog -On`

**Reply by renewtx** [reply]
No -- basically the same (just more up front stuff, but nothing related to the actual field extraction)

pwsh Information: 0 : 2024-12-09 18:14:45.9192	[GetContextAsync]	[0]	[Debug]	Authentication type: AzureADCertificate for scopes https://mysticdev.sharepoint.com/.default	0ms	
pwsh Information: 0 : 2024-12-09 18:14:46.3545	[PnP.Framework]	[0]	[Debug]	Cloning context for https://mysticdev.sharepoint.com/sites/KEEPitTemplateSiteRel2.0	0ms	
pwsh Information: 0 : 2024-12-09 18:14:46.3550	[PnP.Framework]	[0]	[Debug]	Checking for different audience https://mysticdev.sharepoint.com/sites/KEEPitTemplateSiteRel2.0	0ms	
pwsh Information: 0 : 2024-12-09 18:14:46.3709	[GetAccessTokenAsync]	[0]	[Debug]	Authentication type: AzureADCertificate	0ms	
pwsh Information: 0 : 2024-12-09 18:14:47.4038	[Extraction]	[18]	[Debug]	Code execution scope started	0ms	9be3dbfb-e7fa-4dd0-b026-b501b6b860e1
pwsh Information: 0 : 2024-12-09 18:14:47.4040	[Extraction]	[18]	[Debug]	Base template available: GROUP0template	0ms	9be3dbfb-e7fa-4dd0-b026-b501b6b860e1
pwsh Information: 0 : 2024-12-09 18:14:47.4041	[Extraction]	[18]	[Debug]	ProgressDelegate registered	0ms	9be3dbfb-e7fa-4dd0-b026-b501b6b860e1
pwsh Information: 0 : 2024-12-09 18:14:47.4041	[Extraction]	[18]	[Debug]	MessagesDelegate registered	0ms	9be3dbfb-e7fa-4dd0-b026-b501b6b860e1
pwsh Information: 0 : 2024-12-09 18:14:47.4087	[PnP.Framework]	[0]	[Debug]	Cloning context for https://mysticdev.sharepoint.com/sites/KEEPitTemplateSiteRel2.0	0ms	
pwsh Information: 0 : 2024-12-09 18:14:47.4088	[PnP.Framework]	[0]	[Debug]	Checking for different audience https://mysticdev.sharepoint.com/sites/KEEPitTemplateSiteRel2.0	0ms	
pwsh Information: 0 : 2024-12-09 18:14:47.4117	[Fields]	[18]	[Debug]	Code execution scope started	1ms	9be3dbfb-e7fa-4dd0-b026-b501b6b860e1
pwsh Information: 0 : 2024-12-09 18:14:47.9430	[Fields]	[18]	[Debug]	Code execution scope ended	532ms	9be3dbfb-e7fa-4dd0-b026-b501b6b860e1
pwsh Information: 0 : 2024-12-09 18:14:47.9436	[Extraction]	[18]	[Debug]	Code execution scope ended	539ms	9be3dbfb-e7fa-4dd0-b026-b501b6b860e1
pwsh Information: 0 : 2024-12-11 11:32:02.9862	[GetAccessTokenAsync]	[0]	[Debug]	Authentication type: AzureADCertificate	0ms	
pwsh Information: 0 : 2024-12-11 11:32:06.3002	[Extraction]	[18]	[Debug]	Code execution scope started	0ms	25b33cb5-4ef2-4e4f-8d52-0ba9d733e925
pwsh Information: 0 : 2024-12-11 11:32:06.3004	[Extraction]	[18]	[Debug]	Base template available: GROUP0template	0ms	25b33cb5-4ef2-4e4f-8d52-0ba9d733e925
pwsh Information: 0 : 2024-12-11 11:32:06.3004	[Extraction]	[18]	[Debug]	ProgressDelegate registered	0ms	25b33cb5-4ef2-4e4f-8d52-0ba9d733e925
pwsh Information: 0 : 2024-12-11 11:32:06.3004	[Extraction]	[18]	[Debug]	MessagesDelegate registered	0ms	25b33cb5-4ef2-4e4f-8d52-0ba9d733e925
pwsh Information: 0 : 2024-12-11 11:32:06.3077	[PnP.Framework]	[0]	[Debug]	Cloning context for https://mysticdev.sharepoint.com/sites/KEEPitTemplateSiteRel2.0	0ms	
pwsh Information: 0 : 2024-12-11 11:32:06.3078	[PnP.Framework]	[0]	[Debug]	Checking for different audience https://mysticdev.sharepoint.com/sites/KEEPitTemplateSiteRel2.0	0ms	
pwsh Information: 0 : 2024-12-11 11:32:06.3079	[Fields]	[18]	[Debug]	Code execution scope started	0ms	25b33cb5-4ef2-4e4f-8d52-0ba9d733e925
pwsh Information: 0 : 2024-12-11 11:32:06.7801	[Fields]	[18]	[Debug]	Code execution scope ended	472ms	25b33cb5-4ef2-4e4f-8d52-0ba9d733e925
pwsh Information: 0 : 2024-12-11 11:32:06.7803	[Extraction]	[18]	[Debug]	Code execution scope ended	480ms	25b33cb5-4ef2-4e4f-8d52-0ba9d733e925

**Reply by waaromikniet** [reply]
Can you try Set-PnPTraceLog -On -Level debug?
Ah I see it looks already like debug is on

**Reply by renewtx** [reply]
Yes -- that is what I did
![image](https://github.com/user-attachments/assets/546ec555-0d0f-4bbd-a3ee-1c22d02cb28e)

**Reply by waaromikniet** [reply]
I see in the output that the check is done on application permissions and not the delegated ones. I don't see the taxonomy permissions listed in your output. How do you use Connect-PnPOnline?

**Reply by renewtx** [reply]
This is what I use to connect ....

 Connect-PnPOnline -Url "https://mysticdev.sharepoint.com/sites/KEEPitTemplateSiteRel2.0" -ClientId
$clientId -Tenant $tenantId -CertificatePath $certPath -CertificatePassword $certPassword

I set the APP Reference using $clientId = "2ba6c580-dd5a-4a15-9b6f-5c2936eac9a0", which is what is represented in the previous screenshot

![image](https://github.com/user-attachments/assets/f94d4fe3-b048-4a08-bc65-7511cce27fd7)

**Reply by waaromikniet** [reply]
When using a certificate to login you need the Application permissions and not the delegated to my knowledge

![image](https://github.com/user-attachments/assets/c3479685-bf45-4340-a463-e76136a3b951)

**Reply by renewtx** [reply]
Just a quick update ...

Turns out I did have an issue with the app reg in the sense that I had mixed up the variable I should have been using (a little confusion between what I had set on the command line and what i was looking at in my script).  And your suggestions above also helped ... Once I got all that resolved I had more luck.

That said, even after resolving the above I still had the same "Object reference not set" error ... but knowing that all my connection/permissions settings were now synched correctly, I focused on the site I was using as a template -- as mentioned, all the other handlers I was using did work, just not the Fields handler.  So I started looking at my content types and associated columns and I discovered that the formula of one of my (calculated) columns had not published from the content type hub properly ... the formula had lost the actual field references.  Once I fixed that, the handler worked!

Thanks for all your help and clarifications!

**Comment by renewtx** [comment]
This is how I set up the connection 

 Connect-PnPOnline -Url "https://xxxxxxx.sharepoint.com/sites/KEEPitTemplateSiteRel2.0" -ClientId $clientId -Tenant $tenantId -CertificatePath $certPath -CertificatePassword $certPassword

Let me confirm all the variables are set correctly.

Thanks for your responses!
